scripts/apps/dashboard/user-activity/components/UserActivityWidget.tsx

```diff
@@ -6,12 +6,14 @@ import ng from 'core/services/ng';
 import {gettext} from 'core/utils';
 import {SelectUser} from 'core/ui/components/SelectUser';
 import {logger} from 'core/services/logger';
+import {extensions} from 'appConfig';
 
 type FetchFunction = (repo: string, criteria: any) => Promise<any>;
 
 interface IGroup {
     id: string;
     label: string;
+    precondition?: () => boolean;
```

Comment:
Is there a reason you need a precondition? What about using an `if` statement inside `GET_GROUPS`?

```
const GET_GROUPS = () => {
    const markForUserExtensionEnabled = extensions['markForUser'] != null;

    const groups = [
        {
            id: 'created',
            label: gettext('Created by this user'),
            dataSource(fetchFn) {
                return fetchFn('archive', {
                    bool: {
                        must: [
                            // ...
                            {...(markForUserExtensionEnabled ? getMarkedQuery() : {})},
                        ],
                    },
                });
            },
        },
    ];

    if (markForUserExtensionEnabled) {
        groups.push({
            id: 'marked',
            label: gettext('Marked for this user'),
            // ...
        });
    }

    return groups;
};
```

Reply:
I think it's cleaner having a precondition. You could have another group that you don't want to show for example depending on if the user is admin, or something like that. Then I would add all the logic also in GET_GROUPS which IMO is not as clean.

Reply:
You can check for admin rights or permissions using another conditional inside `GET_GROUPS`. How does a precondition make it cleaner?

I'd argue it makes it dirtier. Instead of handling it in a single place, you are now handling it in 2 and have to introduce another member to the interface which you wouldn't need otherwise without no benefits I can see.

Reply:
That check is tight to one group. I'm keeping that group's logic inside the interface. If I add another group with another precondition I just need to add another object, not any code in the GET_GROUPS which is just a function to initialize the groups. Your solution can clutter the GET_GROPS function with logic of each group you add. I don't see any benefits but I see drawbacks.

Reply:
The benefit of using an `if` statement is that the logic is kept in a single place and other parts of the code don't know about it. It translates to being easier to understand and faster to change the code. Not by much too be fair since the file is quite small.

> That check is tight to one group. I'm keeping that group's logic inside the interface.

I get your point about keeping related code together, but it's also kept together when using `ifs`. If you needed to repeat the same if somewhere else, I'd suggest going with the precondition.

> If I add another group with another precondition I just need to add another object, not any code in the GET_GROUPS

I don't see much difference between having to [add an `if` inside the function or an inline function to the group object](https://user-images.githubusercontent.com/2076953/73192409-4678ae00-4129-11ea-98ba-ddb762202d94.png).

> Your solution can clutter the GET_GROPS function with logic of each group you add.

Adding an `if` is 2 lines per group which isn't that different to space that would be required for precondition function.